xray/core/dataset.py

```diff
@@ -1028,6 +1029,58 @@ def sel(self, method=None, **indexers):
         return self.isel(**indexing.remap_label_indexers(self, indexers,
                                                          method=method))
 
+    def isel_points(self, dim='points', **indexers):
+        """Returns a new dataset with each array indexed pointwise along the
+        specified dimension(s).
+
+        This method selects pointwise values from each array and is akin to
+        the NumPy indexing behavior of `arr[[0, 1], [0, 1]]`, except this
+        method does not require knowing the order of each array's dimensions.
+
+        Parameters
+        ----------
+        dim : str, optional
+            Dimension name for which the points will be added to.
+        **indexers : {dim: indexer, ...}
+            Keyword arguments with names matching dimensions and values given
+            by integers, slice objects or arrays. All indexers must be the same
+            length.
+
+        Returns
+        -------
+        obj : Dataset
+            A new Dataset with the same contents as this dataset, except each
+            array and dimension is indexed by the appropriate indexers. In
+            general, each array's data will be a view of the array's data
+            in this dataset, unless numpy fancy indexing was triggered by using
+            an array indexer, in which case the data will be a copy.
+
+        See Also
+        --------
+        Dataset.sel
+        DataArray.isel
+        DataArray.sel
+        DataArray.isel_points
+        """
+        invalid = [k for k in indexers if k not in self.dims]
+        if invalid:
+            raise ValueError("dimensions %r do not exist" % invalid)
+
+        # all the indexers should be iterables
+        keys = indexers.keys()
+        indexers = [(k, ([v] if not isinstance(v, Sequence) else v))
+                    for k, v in iteritems(indexers)]
+
+        # all the indexers should have the same length
+        lengths = set([len(v) for k, v in indexers])
+        if len(lengths) > 1:
+            raise ValueError('All indexers must be the same length')
+
+        return concat([self.isel(**d) for d in
+                       [dict(zip(keys, inds)) for inds in
+                        zip(*[v for k, v in indexers])]],
+                      dim=dim)
```

Comment:
Let's explicitly provide the `coords` and `data_vars` arguments to `concat`:
```python
indexer_dims = set(indexers)

def relevant_keys(mapping):
    return [k for k, v in mapping.items()
            if any(d in indexer_dims for d in v.dims)]

data_vars = relevant_keys(self.data_vars)
coords = relevant_keys(self.coords)
```

This means `concat` doesn't need to look at any data to figure out which variables should be concatenated (vs. variables which were not indexed).

The test case where would be a dataset that aren't has a constant variable, e.g.,
```
ds = xray.Dataset({'x': range(5), 'y': 0})
```

If you do `ds.isel_points(x=[0, 1, 2])`, `y` should still be a scalar.